Implement Hover for Solidity.

Hyperion101010 opened the pull request:
The current implementation contains code for hover and provides
hover for variable types in functions, structs, enums and contract.
Contains tests and documentation(readme) for the hover impl.

Signed-off-by: Shivam Balikondwar <shivambalikondwar@icloud.com>

Commit:
Implement Hover for Solidity.

The current implementation contains code for hover and provides
hover for variable types in functions, structs, enums and contract.
Contains tests and documentation(readme) for the hover impl.

Signed-off-by: Shivam Balikondwar <shivambalikondwar@icloud.com>

seanyoung commented on solang-server/Cargo.toml:
```diff
@@ -12,6 +12,7 @@ bytes = "0.5"
 solang = { git= "https://github.com/hyperledger-labs/solang" }
 dashmap = "3.5.1"
 futures = { version = "0.3", features = ["compat"] }
+inkwell = { version = "0.1.0-llvm10sample", features = ["target-webassembly", "llvm10-0"] }
```
inkwell is not needed; solang depends on inkwell, but solang dependencies will pull that in. solang-server does not depend on inkwell directly.

Hyperion101010 replied:
done

seanyoung commented on solang-server/src/serverutils/backend.rs:
```diff
@@ -12,6 +12,16 @@ use solang::sema::*;
+
+use solang::sema::ast::Expression::*;
+
+use solang::sema::ast::Type;
```
use solang::sema::ast::*; already pulls in 
use solang::sema::ast::Type;
this line is not necessary

Hyperion101010 replied:
done

seanyoung commented on solang-server/src/serverutils/backend.rs:
```diff
@@ -87,6 +97,693 @@ impl Backend {
+        symtab: &sema::symtable::Symtable,
+        ns: &ast::Namespace,
+    ) {
+        match stmt {
```
The params need "constructing" too

seanyoung commented on solang-server/src/serverutils/backend.rs:
```diff
@@ -87,6 +97,693 @@ impl Backend {
+            Statement::Delete(_locs, _typ, expr) => {
+                Backend::construct_expr(expr, lookup_tbl, symtab, ns);
+            }
+            Statement::Destructure(_locs, _vecdestrfield, expr) => {
```
_vecdestrfield needs "constructing" too. e.g.:
```
(a, b) = this_is_function_returning_two_things();
```
Then there should be a hover-over for a and b

seanyoung commented on solang-server/src/serverutils/backend.rs:
```diff
@@ -87,6 +97,693 @@ impl Backend {
+            }
+            Statement::Emit {
+                loc: _,
+                event_no: _,
```
In the code:
``emit Foo(true, 1);`` Foo should have a hoover-over showing the fields of event Foo.

seanyoung commented on solang-server/src/serverutils/backend.rs:
```diff
@@ -87,6 +97,693 @@ impl Backend {
+                catch_param_pos: _,
+                catch_stmt: _,
+            } => {
+                Backend::construct_expr(expr, lookup_tbl, symtab, ns);
```
ok_stmt, catch_stmt, error_stmt also need "Constructing"

seanyoung commented on solang-server/src/serverutils/backend.rs:
```diff
@@ -87,6 +97,693 @@ impl Backend {
+            ConstantVariable(locs, typ, _val1, _val2) => {
+                let typ_msg = Backend::construct_fnc_type(typ, ns);
+
+                let msg = format!("({})", typ_msg);
```
It would be nice if this showed it is a constant, and the constant expression.

seanyoung commented on solang-server/src/serverutils/backend.rs:
```diff
@@ -87,6 +97,693 @@ impl Backend {
+    // Constructs the expression data type messages and stores it in the lookup table
+    fn construct_fnc_type(fnc_typ: &ast::Type, ns: &ast::Namespace) -> String {
+        let msg;
+        match fnc_typ {
```
solang already has a function for this, you can do `fnc_typ.to_string(ns);`

seanyoung commented on solang-server/src/serverutils/backend.rs:
```diff
@@ -87,6 +97,693 @@ impl Backend {
+            Balance(_locs, _typ, expr) => {
+                Backend::construct_expr(expr, lookup_tbl, symtab, ns);
+            }
+            Builtin(_locs, _typ, _builtin, expr) => {
```
The builtins could do with documentation, but solang needs to add those first.

Hyperion101010 replied:
will fix in next eval

seanyoung reviewed (commented):
Looks good!

Commit:
Changes for eval3 review.

Fixed the requested changes. Included traversal for destructure,
try catch and other fixes.

Signed-off-by: Shivam Balikondwar <shivambalikondwar@icloud.com>

Hyperion101010 commented:
Ready for review.
